101-data-share-share-storage-account/azuredeploy.json

```diff
@@ -0,0 +1,190 @@
+      }
+    },
+    "invitationEmail": {
+      "type": "string",
+      "metadata": {
+        "description": "Specify an email address for receiving data share invitations."
+      }
+    },
+    "syncKind": {
+      "type": "string",
+      "defaultValue": "ScheduleBased",
+      "metadata": {
+        "description": "Specify the kind of synchronization."
+      }
+    },
+    "syncInterval": {
+      "type": "string",
+      "defaultValue": "Day",
+      "allowedValues": [ "Day", "Hour" ],
+      "metadata": {
+        "description": "Specify snapshot schedule recurrence."
+      }
+    },
+    "syncTime": {
+      "type": "string",
+      "defaultValue": "[utcNow('yyyy-MM-ddTHH:mm:ssZ')]",
+      "metadata": {
+        "description": "Specify snapshot schedule start time."
+      }
+    }
+  },
+  "variables": {
+    "storageAccountName": "[concat(parameters('projectName'), 'store')]",
+    "containerName": "[concat(parameters('projectName'), 'container')]",
+    "dataShareAccountName": "[concat(parameters('projectName'), 'shareaccount')]",
+    "dataShareName": "[concat(parameters('projectName'),'share')]",
+    "roleAssignmentDeploymentName": "[concat(parameters('projectName'),'roleassignment')]",
+    "roleAssignmentName": "[guid(uniqueString(variables('storageAccountName')))]",
+    "inviteName": "[concat(variables('dataShareName'), 'invite')]",
+    "storageBlobDataReaderRoleDefinitionId": "[resourceId('Microsoft.Authorization/roleDefinitions', '2a2b9908-6ea1-4ae2-8e65-a410df84e7d1')]"
+  },
+  "resources": [
+    {
+      "type": "Microsoft.Storage/storageAccounts",
+      "apiVersion": "2019-06-01",
+      "name": "[variables('storageAccountName')]",
+      "location": "[parameters('location')]",
+      "sku": {
+        "name": "Standard_LRS",
+        "tier": "Standard"
+      },
+      "kind": "StorageV2",
+      "properties": {
+        "accessTier": "Hot"
+      },
+      "resources": [
+        {
+          "type": "blobServices/containers",
+          "apiVersion": "2019-06-01",
+          "name": "[concat('default/', variables('containerName'))]",
+          "dependsOn": [
+            "[variables('storageAccountName')]"
+          ]
+        }
+      ]
+    },
+    {
+      "type": "Microsoft.DataShare/accounts",
+      "apiVersion": "2019-11-01",
+      "name": "[variables('dataShareAccountName')]",
+      "location": "[parameters('location')]",
+      "identity": {
+        "type": "SystemAssigned"
+      },
+      "properties": {}
+    },
+    {
+      "type": "Microsoft.DataShare/accounts/shares",
+      "apiVersion": "2019-11-01",
+      "name": "[concat(variables('dataShareAccountName'), '/', variables('dataShareName'))]",
+      "dependsOn": [
+        "[resourceId('Microsoft.DataShare/accounts', variables('dataShareAccountName'))]"
+      ],
+      "properties": {
+        "shareKind": "CopyBased"
+      }
+    },
+    {
+      "type": "Microsoft.Resources/deployments",
```

Comment:
why is this nested?

Reply:
There are other resources depending on this resource.  Because this is an extension resource, "Microsoft.Storage/storageAccounts/providers/roleAssignments", this is the only way I figured out to specify this resource as the denpendsOn of the dependent resources.

Reply:
You need to use the extensionResourceId function

Reply:
I have made these changes:
...
   {
      "type": "Microsoft.Storage/storageAccounts/providers/roleAssignments",
      "apiVersion": "2020-04-01-preview",
      "name": "[concat(variables('storageAccountName'), '/Microsoft.Authorization/', variables('roleAssignmentName'))]",
      "properties": {
        "roleDefinitionId": "[variables('storageBlobDataReaderRoleDefinitionId')]",
        "principalId": "[reference(resourceId('Microsoft.DataShare/accounts', variables('dataShareAccountName')), '2019-11-01', 'Full').identity.principalId]"
      }
    },
    {
      "type": "Microsoft.DataShare/accounts/shares/dataSets",
      "apiVersion": "2019-11-01",
      "name": "[concat(variables('dataShareAccountName'), '/', variables('dataShareName'),'/', variables('containerName'))]",
      "dependsOn": [
        "[resourceId('Microsoft.DataShare/accounts/shares', variables('dataShareAccountName'), variables('dataShareName'))]",
        "[resourceId('Microsoft.DataShare/accounts', variables('dataShareAccountName'))]",
        "[extensionResourceId(resourceId('Microsoft.Storage/storageAccounts', variables('storageAccountName')), 'Microsoft.Authorization/roleAssignments/', concat(variables('storageAccountName'), '/Microsoft.Authorization/roleAssignments/', variables('roleAssignmentName')))]"
      ],
...

The error message shows "Unable to 
evaluate template language function 'extensionResourceId': function requires exactly two multi-segmented arguments. The first must be the parent resource id while the second must be resource type including resource provider namespace. Current function 
arguments '/subscriptions/65a1016d-0f67-45d2-b838-b8f373d6d52e/resourceGroups/jgaods16rg/providers/Microsoft.Storage/storageAccounts/jgaods16store,Microsoft.Authorization/roleAssignments/,jgaods16store/Microsoft.Authorization/roleAssignments/d1d46494-5223-5
c60-b9e9-c41e7409998c'. "

The document,  https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/template-functions-resource, shows 3 arguments:

extensionResourceId(resourceId, resourceType, resourceName1, [resourceName2], ...)

Is this a bug?

Reply:
Should be:

extensionResourceId(resourceId('Microsoft.Storage/storageAccounts', variables('storageAccountName')), 
 'Microsoft.Authorization/roleAssignments', variables('roleAssignmentName'))